keps/sig-network/20190714-IPv6-beta-proposal.md

```diff
@@ -0,0 +1,95 @@
+---
+title: graduate-ipv6-to-beta
+authors:
+  - "@aojea"
+owning-sig: sig-network
+participating-sigs:
+reviewers:
+  - "@bentheelder"
+  - "@andrewsykim"
+  - "@khenidak"
+approvers:
+  - "@lachie83"
+  - "@thockin"
+editor: TBD
+creation-date: 2019-07-14
+last-updated: 2019-07-23
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Graduate IPv6 to beta
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [User Stories](#user-stories)
+- [Proposal](#proposal)
+- [Design](#design)
+  - [Test Plan](#test-plan)
+    - [Needed Tests](#needed-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+Support for IPv6-only clusters was added in Kubernetes 1.9 as an alpha feature, with version 1.13 the Kubernetes default DNS server changed to CoreDNS which has full IPv6 support, there are several CNI plugins with IPv6 support and Dual Stack support is being implemented during the 1.16 cycle.
+
+
+## Motivation
+
+IPv6 adoption is ramping up with the advances in IoT space and explosion in number of mobile devices, this adoption will continue to grow as we can observe at the Google IPv6 adoption statistics https://www.google.com/intl/en/ipv6/statistics.html.
+
+There are cloud providers that are starting to support IPv6 clusters, the dual stack implementation will facilitate the migration from IPv4 to IPv6 and CI testing was implemented using the kind project.
+
+Therefore, we would like to graduate IPv6 support from Alpha to Beta.
+
+### Goals
+
+* Promote IPv6 to beta version.
+
+### Non-Goals
+
+* IPv6 is NOT Dual-stack  
+
+### User Stories
+
+* A user can deploy, operate and use a kubernetes cluster in an IPv6 only environment.
+
+## Proposal
+
+* Make kind IPv6 e2e jobs mandatory for all PRs
+* Create IPv6 e2e jobs on at least 2 Cloud Providers and publish the results on testgrid
```

Comment:
IMHO this (2 cloud provider IPv6 jobs) is potentially a pretty significant hurdle. 

You can't just create e2e jobs on a provider, you need credits from the provider and a working system to deploy IPv6 clusters on them besides the CI glue.

Today we have some AWS quota and lots of GCP quota, for which we have kops tooling on AWS and a few options on GCE, neither of which sets of (tools, provider) do IPv6 clusters AFAIK.

Is there a concrete plan for this particular point? Otherwise I suspect this will block progress.

Reply:
@BenTheElder we need to shop this around in the sig-cloudprovider and get volunteers to help

Reply:
the main reason for asking this is - since beta means it's on by default, if its on by default but does not work on most of the cloud providers, it's not really useful.

Reply:
Well, technically is on by default, you don't have to enable anything, you configure ipv6 address, it work in ipv6, you configure ipv4 address it works in ipv4.

Reply:
@aojea we don't have a feature flag? hmmm

Reply:
Right, I see that misunderstanding lately, this is not a feature as in typical software development. When you develop your application you usually don't care about the network protocol, what @leblanc and others implemented is that k8s will use ipv6 addressew instead of ipv4, you don't have to turn on anything,

Reply:
@aojea Ack thanks. I am ok with just one cloud provider in that case for beta.